Source/PeleLMTransportProp.cpp

```diff
@@ -1,12 +1,123 @@
+
+   // We shouldn't be here unless we're doing LES
+   AMREX_ALWAYS_ASSERT(m_do_les);
+
+   if (m_les_verbose > 0) {
+     amrex::Print() << "   Computing Turbulent Viscosity with LES model: " << m_les_model
+                    << " for time " << getTime(0, a_time) << std::endl;
+   }
+
+   // Create temporary multifab to store velocity gradient tensor
+   amrex::Vector<amrex::Array<amrex::MultiFab, AMREX_SPACEDIM>> GradVel(finest_level+1);
+   for (int lev = 0; lev <= finest_level; ++lev) {
+     auto ldata_p = getLevelDataPtr(lev,a_time);
+     const auto& ba = ldata_p->state.boxArray();
+     const auto& dm = ldata_p->state.DistributionMap();
+     const auto& factory = ldata_p->state.Factory();
+     constexpr int ncomp = AMREX_SPACEDIM*AMREX_SPACEDIM;
+     for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+       GradVel[lev][idim].define(amrex::convert(ba,IntVect::TheDimensionVector(idim)), dm,
+                                 ncomp, 0, MFInfo(), factory);
+     }
+   }
+
+   // compute the velocity gradient
+   getDiffusionTensorOp()->computeGradientTensor(GetVecOfArrOfPtrs(GradVel),
+                                                 GetVecOfConstPtrs(getVelocityVect(a_time)));
+
+   for (int lev = 0; lev <= finest_level; ++lev) {
+     auto ldata_p = getLevelDataPtr(lev,a_time);
+
+     // Get density and cp (if needed) at faces for computing turbulent transport properties
+     amrex::Array<amrex::MultiFab, AMREX_SPACEDIM> dens_fc;
+     amrex::Array<amrex::MultiFab, AMREX_SPACEDIM> cp_fc;
+     amrex::MultiFab cp_cc;
+     const auto& ba = ldata_p->state.boxArray();
+     const auto& dm = ldata_p->state.DistributionMap();
+     const auto& factory = ldata_p->state.Factory();
+     if (m_incompressible) {
+       // just set density as a constant; don't need to worry about cp
+       for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+         dens_fc[idim].define(amrex::convert(ba,IntVect::TheDimensionVector(idim)), dm,
+                              1, 0, MFInfo(), factory);
+         dens_fc[idim].setVal(m_rho);
+       }
+     } else {
+       // get cp_cc (valid in 1 grow cell for interpolation to FCs)
+       int ngrow = 1;
+       cp_cc.define(ba, dm, 1, ngrow, MFInfo(), factory);
+       auto const& state_arr      = ldata_p->state.const_arrays();
+       auto const& cp_arr       = cp_cc.arrays();
+       amrex::ParallelFor(cp_cc, cp_cc.nGrowVect(), [=]
+                          AMREX_GPU_DEVICE (int box_no, int i, int j, int k) noexcept
+                          {
+                            getCpmixGivenRYT( i, j, k,
+                                              Array4<Real const>(state_arr[box_no], DENSITY),
+                                              Array4<Real const>(state_arr[box_no], FIRSTSPEC),
+                                              Array4<Real const>(state_arr[box_no], TEMP),
+                                              Array4<Real      >(cp_arr[box_no]) );
+                          });
+
```

Comment:
You'll need to add a Gpu::streamSynchronize() at this point because the multiarrays ParallelFor are non-blocking.